dra-evolution/README.md

```diff
@@ -118,36 +135,43 @@ spec:
       requests:
         cpu: 10m
         memory: 10Mi
-    devices:
-    - name: gpu
-  deviceClaims:
-    devices:
-    - name: gpu
-      class: example.com-foozer
+      devices:
+      - claimName: gpu
+        deviceName: gpu-one
+  - image: registry.k8s.io/pause:3.6
+    name: my-container
+    resources:
+      requests:
+        cpu: 10m
+        memory: 10Mi
+      devices:
+      - claimName: gpu
+        deviceName: gpu-two
+  resourceClaims:
+  - name: gpu
+    source:
+      resourceClaimTemplate: two-foozers
 ```
 
-Each entry in `spec.deviceClaims.devices` is just a name/class pair, but in fact
-serves as a template to generate claims that exist with the lifecycle of the
-pod. We may want to add `ObjectMeta` here as well, since it is behaving as a
-template, to allow setting labels, etc.
-
-The second form of `DeviceClaims` is a single struct with an ObjectMeta, and a
-claim name. The key with this form is that it is not *list* of named objects.
-Instead, it is a reference to a single claim object, and the named entries are
-*inside* the referenced object. This is to avoid a two-key mount in the
-`spec.containers[*].devices` entry. If that's not important, then we can tweak
-this a bit. In any case, this form allows claims which follow the lifecycle of
-the pod, similar to the first form. Since a top-level API claim spec can can
-contain multiple claim instances, this should be equally as expressive as if we
-included `matchAttributes` in the `PodSpec`, without having to do so.
-
-The third form of `DeviceClaims` is just a string; it is a claim name and allows
-the user to share a pre-provisioned claim between pods.
-
-Given that the first and second forms both have a template-like structure, we
-may want to combine them and use two-key indexing in the mounts. If we do so, we
-still want the direct specification of the class, so that the most common case
-does not need separate object just to reference a class.
+Resource classes are capable of describing everything that a user might put
+into a claim. Therefore a simple claim or claim template might contain nothing
+but a resource class name. For this simple case, a new `claimWithClassName` gets
+added which creates such a claim. Here object meta is supported:
+
+```yaml
+  resourceClaims:
+  - name: gpu
+    source:
+      forClass:
```

Comment:
Why `for` and not `from` -- to me it feels like we build the claims 'from' classes, not 'for' classes?

Also, the alternative fields here are `resourceClaimName` and `resourceClaimTemplateName`, so should we keep `resource*` in the name of the field to be symmetrical,  i.e. putting it all together:
```
resourceClaims:
  - name: gpu
    source:
      fromResourceClass:
        className: two-foozers-class
        metdadata:
          labels:
            foo: bar
```

Also, it's unclear to me what the embedded `metadata` section does. Is it metadata that will be added to the claim that is generated from this class?

Reply:
Also, regarding our discussion offline about the new `classes` field in claim.spec.requests`. If we want to support that use case, do we also need to support listing multiple classes here as well?

```
resourceClaims:
  - name: gpu
    source:
      fromResourceClasses:
        classNames:
        - two-foozers-class
        metdadata:
          labels:
            foo: bar
```

Reply:
And if we do, should we make the structure in the clam look somewhat symmetrical:
```
  requests:
  - fromResourceClasses:
      classNames:
      - gpu.nvidia.com
      requirements:
      - resource:
          name: memory
          quantity: 80Gi
```

Reply:
> Also, it's unclear to me what the embedded metadata section does. Is it metadata that will be added to the claim that is generated from this class?

Yes.

> fromResourceClasses:

"From" doesn't sound right to me either because the claim is not created "from" the class. It's an independent entity that references the class. How about "createClaim"?

```yaml
  resourceClaims:
  - name: gpu
    source:
       createClaim:
          inherit:
          - resourceClass: gpu.nvidia.com
          metdadata:
            labels:
              foo: bar
  - name: nic
    source:
      resourceClaimTemplateName: nic-claim-template
```

Reply:
Something still doesn't sit well with me for all of this. I can't quite put my finger on it, but it "feels" inconsistent and non-intuitive to me.

For example:
* When using a template we also create a claim -- does that not belong under `createClaim` now too then?
* Why do we actually even need `source`? Are there any other fields we ever expect to be sitting underneath `source`, or does it just exist to encapsulate a set of one-ofs?
* Does it make sense to just drop the "shorthand" for referencing classes here altogether and just force people to create a template (that can then list a set of classes if desired)? Given that we plan to ultimately inline templates anyway, I see no reason to add this complexity now.

Meaning we simply go back to what we had before: 
```
  resourceClaims:
  - name: gpu
      source:
        resourceClaimName: gpu-claim
```
or
```
  resourceClaims:
  - name: gpu
      source:
        resourceClaimTemplateName: gpu-claim-template
```
Or we drop the `source`:
```
  resourceClaims:
  - name: gpu
     resourceClaimName: gpu-claim
```
or
```
  resourceClaims:
  - name: gpu
     resourceClaimTemplateName: gpu-claim-template
```
Where in the future this becomes the following once pulled into core/v1:
```
  resourceClaims:
  - name: gpu
     resourceClaimName: gpu-claim
```
or
```
  resourceClaims:
  - name: gpu
     template:
       spec:
         ...
```

Reply:
> When using a template we also create a claim -- does that not belong under createClaim now too then?

It's an alternative in the `source` one-of, so it cannot be there, even if the naming would imply it.

> Why do we actually even need source? 

I think we can drop it and simply inline the one-of [ClaimSource](https://github.com/kubernetes-sigs/wg-device-management/blob/c920b2be9077bdb0d0ba6369e1d6d0f89239982c/dra-evolution/pkg/api/pod_types.go#L206-L210). We've done that everywhere else in this prototype. It's then not consistent with the volume API, but I think we are past that now (see also my open question in https://github.com/kubernetes-sigs/wg-device-management/pull/14#discussion_r1607867479).

If we do that, the YAML becomes:
```yaml
 resourceClaims:
  - name: gpu
    createClaim:
        inherit:
        - resourceClass: gpu.nvidia.com
        metdadata:
          labels:
            foo: bar
  - name: nic
    resourceClaimTemplateName: nic-claim-template
```

> Does it make sense to just drop the "shorthand" for referencing classes here altogether and just force people to create a template (that can then list a set of classes if desired)? 

That would be my preference. I added this shorthand because [k8srm-prototype](https://github.com/kubernetes-sigs/wg-device-management/tree/c920b2be9077bdb0d0ba6369e1d6d0f89239982c/k8srm-prototype) had it, not because I'm really a fan of it myself.

Reply:
If we drop the `Name` suffix, it becomes:
```yaml
resourceClaims:
  - name: gpu
    createClaim:
      inherit:
      - resourceClass: gpu.nvidia.com
      metdadata:
        labels:
          foo: bar
  - name: nic
    resourceClaimTemplate: nic-claim-template
```

Reply:
I am also not really understanding this hunk.  What problem is this addressing, @pohly?

Reply:
This corresponds to https://github.com/kubernetes-sigs/wg-device-management/blob/c920b2be9077bdb0d0ba6369e1d6d0f89239982c/k8srm-prototype/testdata/pod-one-container-one-gpu.yaml#L19-L22: instead of creating a claim or claim template and referencing those, users name a resource class and Kubernetes creates a claim that uses that class.

The exact same thing can be done with a claim template. It's just syntactic sugar that was added to k8srm-prototype to make the UX simpler. Shall I simply remove it from this PR?

Before I do that and then close this discussion thread, do we agree to drop the `Name` suffix from fields?

Reply:
The convention for "references" is:
* If the Kind is effectively hardcoded, you can use a `<kind>Name string` field
* If the Kind is not hardcoded, you can used `<something>Ref <reftype>` field, where the reftype is a struct of api-group + kind or resource + name.

I think, for the sake of progress you should check-in changes here quickly, so we can iterate on small decisions.  I feel like we're going to end up doing a top-to-bottom API review anyway.  The little things don't matter yet.

Reply:
I've removed the `forClass` syntactic sugar and switched to `resourceClassName` instead of `name` in the `ClassReference`.

Reply:
I'm not sure I follow. Are you now talking about pod spec changes?

Reply:
Yeah, we've always talked about eventually embedding ResourceClaimTemplates directly in the pod spec and removing the need for the top level object. I'm just asking what it will look like once we support such embedding.

Reply:
Got it.

It would be something like this:
```
resourceClaims:
- resourceClaimTemplateName: my-claim-template
- resourceClaimTemplate:
  metadata:
    labels:
      my-label-foo: bar
  spec:
    requests:
    - resourceClassName: example.com-foozer
```

Of course, the actual `resourceClaimTemplate` field name is up for debate. We can also make it shorter, for example with `createClaim`.

Reply:
Probably just `template`, since it's already under `resourceClaims`.

Reply:
I had forgotten that this is in the v1 core API. Removing `source:` field via inlining the one-of is a breaking change, albeit of an alpha field. Is it still worth it? I think it is.

I verified that this is (so far - some tests still running) not causing problems: https://github.com/kubernetes/kubernetes/pull/125116

Reply:
We decided to go ahead with simplifying the v1 pod API because now we can still do it (alpha).